stan/math/prim/mat/prob/ordered_logistic_lpmf.hpp

```diff
@@ -112,126 +50,105 @@ typename return_type<T_loc, T_cut>::type ordered_logistic_lpmf(
+    int size_c_new = c_vec[i].size();
+
+    check_size_match(function, "Size of one of the vectors of cutpoints ",
+                     size_c_new, "Size of another vector of the cutpoints ",
+                     size_c_old);
+    size_c_old = size_c_new;
   }
-  return logp_n;
-}
 
-template <typename T_loc, typename T_cut>
-typename return_type<T_loc, T_cut>::type ordered_logistic_lpmf(
-    const std::vector<int>& y,
-    const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
-    const Eigen::Matrix<T_cut, Eigen::Dynamic, 1>& c) {
-  return ordered_logistic_lpmf<false>(y, lambda, c);
-}
+  for (int n = 0; n < N; n++) {
+    check_bounded(function, "Random variable", y_vec[n], 1, K);
+    check_finite(function, "Location parameter", lam_vec[n]);
+  }
 
-/**
- * Returns the (natural) log probability of the specified array
- * of integers given the vector of continuous locations and
- * array of specified cutpoints in an ordered logistic model.
- *
- * <p>Typically the continous location
- * will be the dot product of a vector of regression coefficients
- * and a vector of predictors for the outcome.
- *
- * @tparam propto True if calculating up to a proportion.
- * @tparam T_y Type of y variable (should be std::vector<int>).
- * @tparam T_loc Location type.
- * @tparam T_cut Cut-point type.
- * @param y Array of integers
- * @param lambda Vector of continuous location variables.
- * @param c array of Positive increasing vectors of cutpoints.
- * @return Log probability of outcome given location and
- * cutpoints.
- * @throw std::domain_error If the outcome is not between 1 and
- * the number of cutpoints plus 2; if the cutpoint vector is
- * empty; if the cutpoint vector contains a non-positive,
- * non-finite value; or if the cutpoint vector is not sorted in
- * ascending order.
- * @throw std::invalid_argument If y and lambda are different
- * lengths, or if y and the array of cutpoints are of different
- * lengths.
- */
-template <bool propto, typename T_loc, typename T_cut>
-typename return_type<T_loc, T_cut>::type ordered_logistic_lpmf(
-    const std::vector<int>& y,
-    const Eigen::Matrix<T_loc, Eigen::Dynamic, 1>& lambda,
-    const std::vector<Eigen::Matrix<T_cut, Eigen::Dynamic, 1> >& c) {
-  typename return_type<T_loc, T_cut>::type logp_n(0.0);
+  for (size_t i = 0, size_ = C_l; i < size_; i++) {
+    check_ordered(function, "Cut-points", c_vec[i]);
+    check_greater(function, "Size of cut points parameter", c_vec[i].size(), 0);
+    check_finite(function, "Final cut-point", c_vec[i](c_vec[i].size() - 1));
+    check_finite(function, "First cut-point", c_vec[i](0));
+  }
 
-  using std::exp;
-  using std::log;
+  operands_and_partials<T_loc, T_cut> ops_partials(lambda, c);
 
-  static const char* function = "ordered_logistic";
+  T_partials_return logp(0.0);
+  T_partials_vec c_dbl = value_of(c_vec[0]).template cast<T_partials_return>();
 
-  int N = lambda.size();
+  for (int n = 0; n < N; ++n) {
+    if (C_l > 1)
```

Comment:
The vector_seq_view should prevent the need for this check: https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/meta/vector_seq_view.hpp#L40

Reply:
This, and the code in the other vector_seq_view comment, are there to reduce unnnecessary copying. So this code:
```
  T_partials_vec c_dbl = value_of(c_vec[0]).template cast<T_partials_return>();

  for (int n = 0; n < N; ++n) {
    if (C_l > 1)
      c_dbl = value_of(c_vec[n]).template cast<T_partials_return>();
```
If only a single vector of cutpoints is provided, it only gets copied into ```c_dbl``` once. Whereas moving the declaration into the loop would mean that the same vector gets copied ```N``` times.